File: packages/protocol/src/validation.ts
import { protocolErrorCodes, SpecterProtocolError } from './errors'
import { digestSpecification, parseSpecification } from '@specter-ts/spec'
import {
  observationKinds,
  SPECTER_PROTOCOL_VERSION,
  type JsonValue,
  type ProtocolMessage,
  type RuntimeObservationBatch,
  type RuntimeObservation,
  type SpecificationPublication,
} from './types'

export function assertJsonValue(
  value: unknown,
  path = '$',
  seen = new Set<object>(),
): asserts value is JsonValue {
  if (value === null || typeof value === 'string' || typeof value === 'boolean')
    return
  if (typeof value === 'number') {
    if (
      !Number.isFinite(value) ||
      (!Number.isSafeInteger(value) && Number.isInteger(value))
    ) {
      fail(`${path} must be a finite JSON-safe number`)
    }
    return
  }
  if (typeof value !== 'object') fail(`${path} is not JSON-compatible`)
  if (seen.has(value)) fail(`${path} contains a cycle`)
  seen.add(value)
  try {
    if (Array.isArray(value)) {
      value.forEach((item, index) => {
        assertJsonValue(item, `${path}[${index}]`, seen)
      })
      return
    }
    const prototype = Object.getPrototypeOf(value)
    if (prototype !== Object.prototype && prototype !== null)
      fail(`${path} must be a plain object`)
    for (const [key, item] of Object.entries(value))
      assertJsonValue(item, `${path}.${key}`, seen)
  } finally {
    seen.delete(value)
  }
}

export function parseProtocolMessage(value: unknown): ProtocolMessage {
  assertJsonValue(value)
  const message = record(value, '$')
  integer(message.protocolVersion, '$.protocolVersion')
  if (message.protocolVersion !== SPECTER_PROTOCOL_VERSION) {
    throw new SpecterProtocolError({
      code: protocolErrorCodes.versionMismatch,
      message: `Protocol major ${String(message.protocolVersion)} is unsupported; expected ${SPECTER_PROTOCOL_VERSION}.`,
      status: 426,
      details: {
        expected: SPECTER_PROTOCOL_VERSION,
        received: message.protocolVersion as number,
      },
    })
  }
  const kind = string(message.kind, '$.kind')
  string(message.requestId, '$.requestId')

  switch (kind) {
    case 'observations.batch': {
      const observations = array(message.observations, '$.observations')
      if (observations.length > 100)
        fail('$.observations must contain at most 100 items')
      observations.forEach((observation, index) => {
        runtimeObservation(observation, `$.observations[${index}]`)
      })
      break
    }
    case 'observations.ack':
      integer(message.accepted, '$.accepted')
      integer(message.duplicates, '$.duplicates')
      optionalUniqueStrings(
        message.rejectedObservationIds,
        '$.rejectedObservationIds',
      )
      break
    case 'specifications.publish': {
      runtimeSource(message.source, '$.source')
      const specifications = array(message.specifications, '$.specifications')
      if (specifications.length > 100)
        fail('$.specifications must contain at most 100 items')
      specifications.forEach((value, index) => {
        const path = `$.specifications[${index}]`
        const published = record(value, path)
        const digest = specificationDigest(published.digest, `${path}.digest`)
        let canonicalDigest: string
        try {
          canonicalDigest = digestSpecification(
            parseSpecification(published.document),
          )
        } catch (cause) {
          throw new SpecterProtocolError({
            code: protocolErrorCodes.invalidMessage,
            message: `${path}.document is not a valid Slice specification.`,
            details: {
              path: `${path}.document`,
              reason: cause instanceof Error ? cause.message : String(cause),
            },
            cause,
          })
        }
        if (canonicalDigest !== digest)
          fail(
            `${path}.digest does not match the canonical specification document`,
          )
      })
      break
    }
    case 'specifications.ack':
      specificationDigests(message.acceptedDigests, '$.acceptedDigests')
      if (message.rejectedDigests !== undefined)
        specificationDigests(message.rejectedDigests, '$.rejectedDigests')
      break
    default:
      fail(`$.kind has unknown message kind ${kind}`)
  }
  return value as ProtocolMessage
}

export function parseProtocolJson(input: string): ProtocolMessage {
  let value: unknown
  try {
    value = JSON.parse(input)
  } catch (cause) {
    throw new SpecterProtocolError({
      code: protocolErrorCodes.invalidJson,
      message: 'Malformed JSON message.',
      cause,
    })
  }
  return parseProtocolMessage(value)
}

export function parseRuntimeObservation(value: unknown): RuntimeObservation {
  assertJsonValue(value)
  runtimeObservation(value, '$')
  return value as RuntimeObservation
}

export function assertProtocolMessage(
  value: unknown,
): asserts value is ProtocolMessage {
  parseProtocolMessage(value)
}

export function parseRuntimeObservationBatch(
  value: unknown,
): RuntimeObservationBatch {
  assertRuntimeObservationBatch(value)
  return value
}

export function assertRuntimeObservationBatch(
  value: unknown,
): asserts value is RuntimeObservationBatch {
  const parsed = parseProtocolMessage(value)
  if (parsed.kind !== 'observations.batch') {
    fail('$.kind must be observations.batch')
  }
}

export function parseSpecificationPublication(
  value: unknown,
): SpecificationPublication {
  assertSpecificationPublication(value)
  return value
}

export function assertSpecificationPublication(
  value: unknown,
): asserts value is SpecificationPublication {
  const parsed = parseProtocolMessage(value)
  if (parsed.kind !== 'specifications.publish')
    fail('$.kind must be specifications.publish')
}

function runtimeObservation(value: unknown, path: string) {
  const input = record(value, path)
  causality(input, path)
  string(input.observationId, `${path}.observationId`)
  integer(input.sequence, `${path}.sequence`)
  timestamp(input.observedAt, `${path}.observedAt`)
  enumeration(input.kind, observationKinds, `${path}.kind`)
  runtimeSource(input.source, `${path}.source`)
  if (input.events !== undefined) events(input.events, `${path}.events`)
  if (input.error !== undefined) structuredError(input.error, `${path}.error`)
  optionalString(input.commandType, `${path}.commandType`)
  optionalString(input.queryType, `${path}.queryType`)
  optionalString(input.slice, `${path}.slice`)
  if (input.specificationDigest !== undefined)
    specificationDigest(
      input.specificationDigest,
      `${path}.specificationDigest`,
    )
  optionalString(input.reaction, `${path}.reaction`)
  if (input.outcome !== undefined)
    enumeration(
      input.outcome,
      ['succeeded', 'rejected', 'failed'],
      `${path}.outcome`,
    )
  if (input.attributes !== undefined)
    record(input.attributes, `${path}.attributes`)
  optionalInteger(input.cursor, `${path}.cursor`)
  optionalInteger(input.droppedCount, `${path}.droppedCount`)
}

function runtimeSource(value: unknown, path: string) {
  const source = record(value, path)
  for (const key of [
    'application',
    'environment',
    'runtimeLanguage',
    'runtimeVersion',
    'instanceId',
    'eventLogId',
  ]) {
    string(source[key], `${path}.${key}`)
  }
}

function causality(input: Record<string, unknown>, path = '$') {
  string(input.operationId, `${path}.operationId`)
  optionalString(input.correlationId, `${path}.correlationId`)
  optionalUniqueStrings(input.parentOperationIds, `${path}.parentOperationIds`)
  optionalUniqueStrings(input.triggeringEventIds, `${path}.triggeringEventIds`)
  optionalString(input.reactionPassId, `${path}.reactionPassId`)
  optionalString(input.deliveryId, `${path}.deliveryId`)
  optionalString(input.attemptId, `${path}.attemptId`)
  if (input.triggeringEventOrder !== undefined) {
    const range = record(
      input.triggeringEventOrder,
      `${path}.triggeringEventOrder`,
    )
    integer(range.from, `${path}.triggeringEventOrder.from`)
    integer(range.to, `${path}.triggeringEventOrder.to`)
    if ((range.to as number) < (range.from as number))
      fail(`${path}.triggeringEventOrder.to must be >= from`)
  }
}

function events(value: unknown, path: string) {
  let previousOrder = -1
  array(value, path).forEach((item, index) => {
    const event = record(item, `${path}[${index}]`)
    string(event.eventId, `${path}[${index}].eventId`)
    string(event.type, `${path}[${index}].type`)
    integer(event.order, `${path}[${index}].order`)
    if ((event.order as number) <= previousOrder)
      fail(`${path} must be strictly ascending by Event order`)
    previousOrder = event.order as number
    integer(event.commitVersion, `${path}[${index}].commitVersion`)
    timestamp(event.recordedAt, `${path}[${index}].recordedAt`)
    if (event.attributes !== undefined)
      record(event.attributes, `${path}[${index}].attributes`)
  })
}

function structuredError(value: unknown, path: string) {
  const error = record(value, path)
  string(error.code, `${path}.code`)
  string(error.message, `${path}.message`)
  if (error.details !== undefined)
    assertJsonValue(error.details, `${path}.details`)
  if (error.retryable !== undefined && typeof error.retryable !== 'boolean')
    fail(`${path}.retryable must be a boolean`)
}

function record(value: unknown, path: string): Record<string, unknown> {
  if (typeof value !== 'object' || value === null || Array.isArray(value))
    fail(`${path} must be an object`)
  return value as Record<string, unknown>
}
function array(value: unknown, path: string): readonly unknown[] {
  if (!Array.isArray(value)) fail(`${path} must be an array`)
  return value
}
function string(value: unknown, path: string): string {
  if (typeof value !== 'string' || value.length === 0)
    fail(`${path} must be a non-empty string`)
  return value
}
function optionalString(value: unknown, path: string) {
  if (value !== undefined) string(value, path)
}
function optionalUniqueStrings(value: unknown, path: string) {
  if (value === undefined) return
  const items = array(value, path)
  const unique = new Set<string>()
  items.forEach((item, index) => {
    const parsed = string(item, `${path}[${index}]`)
    if (unique.has(parsed)) fail(`${path} must contain unique strings`)
    unique.add(parsed)
  })
}
function specificationDigest(value: unknown, path: string): `sha256:${string}` {
  const digest = string(value, path)
  if (!/^sha256:[a-f0-9]{64}$/.test(digest))
    fail(`${path} must be a canonical sha256 digest`)
  return digest as `sha256:${string}`
}
function specificationDigests(value: unknown, path: string) {
  const digests = array(value, path)
  const unique = new Set<string>()
  digests.forEach((value, index) => {
    const digest = specificationDigest(value, `${path}[${index}]`)
    if (unique.has(digest)) fail(`${path} must contain unique digests`)
    unique.add(digest)
  })
}
function integer(value: unknown, path: string) {
  if (!Number.isSafeInteger(value) || (value as number) < 0)
    fail(`${path} must be a non-negative JSON-safe integer`)
}
function optionalInteger(value: unknown, path: string) {
  if (value !== undefined) integer(value, path)
}
function timestamp(value: unknown, path: string) {
  const input = string(value, path)
  const match =
    /^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})(?:\.\d+)?Z$/.exec(input)
  if (!match) fail(`${path} must be an RFC 3339 UTC timestamp`)
  const [
    ,
    yearInput,
    monthInput,
    dayInput,
    hourInput,
    minuteInput,
    secondInput,
  ] = match
  const year = Number(yearInput)
  const month = Number(monthInput)
  const day = Number(dayInput)
  const hour = Number(hourInput)
  const minute = Number(minuteInput)
  const second = Number(secondInput)
  const daysInMonth = [
    31,
    year % 4 === 0 && (year % 100 !== 0 || year % 400 === 0) ? 29 : 28,
    31,
    30,
    31,
    30,
    31,
    31,
    30,
    31,
    30,
    31,
  ]
  if (
    month < 1 ||
    month > 12 ||
    day < 1 ||
    day > (daysInMonth[month - 1] ?? 0) ||
    hour > 23 ||
    minute > 59 ||
    second > 59
  ) {
    fail(`${path} must be an RFC 3339 UTC timestamp`)
  }
}
function enumeration(value: unknown, allowed: readonly string[], path: string) {
  if (typeof value !== 'string' || !allowed.includes(value))
    fail(`${path} must be one of ${allowed.join(', ')}`)
}
function fail(message: string): never {
  throw new SpecterProtocolError({
    code: protocolErrorCodes.invalidMessage,
    message,
  })
}

File: runtimes/go/protocol/validation.go
package protocol

import (
	"encoding/json"
	"errors"
	"fmt"
	"math"
	"math/big"
	"strings"

	"github.com/devagrawal09/specter/runtimes/go/specter"
)

const maxSafeInteger int64 = 9_007_199_254_740_991

var observationKinds = map[string]struct{}{
	"command.started": {}, "command.completed": {}, "command.rejected": {}, "command.failed": {},
	"query.started": {}, "query.completed": {}, "query.rejected": {}, "query.failed": {},
	"events.persisted":       {},
	"slice.catch-up.started": {}, "slice.catch-up.completed": {}, "slice.catch-up.failed": {},
	"subscription.invalidated": {},
	"replay.started":           {}, "replay.completed": {}, "replay.failed": {},
	"reaction.pass.started": {}, "reaction.pass.completed": {}, "reaction.pass.failed": {},
	"reaction.run.started": {}, "reaction.run.completed": {}, "reaction.run.failed": {},
	"outbox.enqueued": {}, "outbox.attempted": {}, "outbox.retry-scheduled": {}, "outbox.dead-lettered": {},
	"telemetry.dropped": {},
}

// ValidateMessageJSON validates either message family in the observation
// protocol while ignoring unknown optional fields.
func ValidateMessageJSON(data []byte) error {
	if !json.Valid(data) {
		return &specter.Error{Code: specter.ErrInvalidJSON, Message: "Malformed JSON message."}
	}
	decoder := json.NewDecoder(strings.NewReader(string(data)))
	decoder.UseNumber()
	var jsonValue any
	if decoder.Decode(&jsonValue) != nil || !validateJSONValue(jsonValue) {
		return invalidMessage("Protocol messages must contain JSON-safe values.")
	}
	var fields map[string]json.RawMessage
	if err := json.Unmarshal(data, &fields); err != nil || fields == nil {
		return invalidMessage("Protocol message must be an object.")
	}
	if !rawInteger(fields, "protocolVersion", true) {
		return invalidMessage("protocolVersion must be a non-negative JSON-safe integer.")
	}
	var envelope Envelope
	if err := json.Unmarshal(data, &envelope); err != nil {
		return invalidMessage("Protocol envelope fields do not match the protocol types.")
	}
	switch envelope.Kind {
	case "observations.batch":
		var batch RuntimeObservationBatch
		if err := json.Unmarshal(data, &batch); err != nil {
			return invalidMessage("Observation batch fields do not match the protocol types.")
		}
		return ValidateObservationBatch(batch)
	case "observations.ack":
		if failure := validateEnvelope(envelope, "observations.ack"); failure != nil {
			return failure
		}
		var acknowledgement ObservationAcknowledgement
		if err := json.Unmarshal(data, &acknowledgement); err != nil {
			return invalidMessage("Observation acknowledgement fields do not match the protocol types.")
		}
		return validateObservationAcknowledgementShape(acknowledgement, fields)
	default:
		return invalidMessage("Unknown observation protocol message kind.")
	}
}

// ValidateObservationBatch checks the language-neutral wire contract before a
// producer sends a batch. Unknown optional JSON fields are intentionally ignored.
func ValidateObservationBatch(batch RuntimeObservationBatch) error {
	if failure := validateEnvelope(batch.Envelope, "observations.batch"); failure != nil {
		return failure
	}
	if batch.decoded && !batch.fieldsValid {
		return invalidMessage("observations must be present and contain an array.")
	}
	if batch.Observations == nil {
		return invalidMessage("observations must contain an array.")
	}
	if len(batch.Observations) > 100 {
		return invalidMessage("Observation batches may contain at most 100 observations.")
	}
	for _, observation := range batch.Observations {
		if failure := validateObservation(observation); failure != nil {
			return failure
		}
	}
	return nil
}

func validateObservation(observation RuntimeObservation) *specter.Error {
	if observation.ObservationID == "" || observation.OperationID == "" || !safeInteger(observation.Sequence) || observation.ObservedAt.IsZero() || observation.Kind == "" {
		return invalidMessage("Observation identity, safe sequence, timestamp, kind, and operationId are required.")
	}
	if _, offset := observation.ObservedAt.Zone(); offset != 0 {
		return invalidMessage("Observation timestamp must be UTC.")
	}
	if observation.decoded && (!observation.fieldsValid || !observation.timestampValid || !observation.attributesValid || !observation.errorValid) {
		return invalidMessage("Observation fields do not match the protocol types.")
	}
	if _, ok := observationKinds[observation.Kind]; !ok {
		return invalidMessage("Unknown runtime observation kind.")
	}
	source := observation.Source
	if source.Application == "" || source.Environment == "" || source.RuntimeLanguage == "" || source.RuntimeVersion == "" || source.InstanceID == "" || source.EventLogID == "" {
		return invalidMessage("Observation source fields are required.")
	}
	if !uniqueNonempty(observation.ParentOperationIDs) || !uniqueNonempty(observation.TriggeringEventIDs) {
		return invalidMessage("Observation causal IDs must be nonempty and unique.")
	}
	if order := observation.TriggeringEventOrder; order != nil && (!safeInteger(order.From) || !safeInteger(order.To) || order.To < order.From) {
		return invalidMessage("Observation triggering Event order range is invalid.")
	}
	if !safeInteger(observation.Cursor) || !safeInteger(observation.DroppedCount) {
		return invalidMessage("Observation cursor and dropped count must be safe integers.")
	}
	previousEventOrder := int64(-1)
	for _, event := range observation.Events {
		if event.EventID == "" || event.Type == "" || event.RecordedAt.IsZero() || !safeInteger(event.Order) || !safeInteger(event.CommitVersion) {
			return invalidMessage("Observation Event references require identity, UTC timestamp, and safe order and version.")
		}
		if int64(event.Order) <= previousEventOrder {
			return invalidMessage("Observation Event references must be strictly ascending by Event order.")
		}
		previousEventOrder = int64(event.Order)
		if _, offset := event.RecordedAt.Zone(); offset != 0 {
			return invalidMessage("Observation Event timestamps must be UTC.")
		}
		if event.decoded && (!event.fieldsValid || !event.timestampValid || !event.attributesValid) {
			return invalidMessage("Observation Event fields do not match the protocol types.")
		}
		if event.Attributes != nil && !validateJSONValue(event.Attributes) {
			return invalidMessage("Observation Event attributes must contain JSON values with safe integers.")
		}
	}
	if observation.Error != nil && (observation.Error.Code == "" || observation.Error.Message == "" || !validateJSONValue(observation.Error.Details)) {
		return invalidMessage("Observation errors require a code and message.")
	}
	if observation.Attributes != nil && !validateJSONValue(observation.Attributes) {
		return invalidMessage("Observation attributes must contain JSON values with safe integers.")
	}
	if observation.Outcome != "" && observation.Outcome != "succeeded" && observation.Outcome != "rejected" && observation.Outcome != "failed" {
		return invalidMessage("Observation outcome is invalid.")
	}
	return nil
}

func validateObservationAcknowledgement(batch RuntimeObservationBatch, acknowledgement ObservationAcknowledgement, fields map[string]json.RawMessage) error {
	if err := validateObservationAcknowledgementShape(acknowledgement, fields); err != nil {
		return err
	}
	observationIDs := make(map[string]struct{}, len(batch.Observations))
	for _, observation := range batch.Observations {
		observationIDs[observation.ObservationID] = struct{}{}
	}
	rejected := make(map[string]struct{}, len(acknowledgement.RejectedObservationIDs))
	for _, id := range acknowledgement.RejectedObservationIDs {
		if _, exists := observationIDs[id]; !exists {
			return fmt.Errorf("specter protocol: acknowledgement rejected an observation outside the batch")
		}
		rejected[id] = struct{}{}
	}
	accounted := int(acknowledgement.Accepted) + int(acknowledgement.Duplicates) + len(rejected)
	if accounted != len(batch.Observations) {
		return fmt.Errorf("specter protocol: acknowledgement accounts for %d of %d observations", accounted, len(batch.Observations))
	}
	return nil
}

func validateObservationAcknowledgementShape(acknowledgement ObservationAcknowledgement, fields map[string]json.RawMessage) error {
	if !hasFields(fields, "accepted", "duplicates") || !rawInteger(fields, "accepted", true) || !rawInteger(fields, "duplicates", true) || !safeInteger(acknowledgement.Accepted) || !safeInteger(acknowledgement.Duplicates) {
		return invalidMessage("Observation acknowledgement counts must be non-negative safe integers.")
	}
	if raw, present := fields["rejectedObservationIds"]; present && !rawArray(raw) {
		return invalidMessage("rejectedObservationIds must be an array when present.")
	}
	rejected := make(map[string]struct{}, len(acknowledgement.RejectedObservationIDs))
	for _, id := range acknowledgement.RejectedObservationIDs {
		if id == "" {
			return invalidMessage("Rejected observation IDs must be nonempty.")
		}
		if _, duplicate := rejected[id]; duplicate {
			return invalidMessage("Rejected observation IDs must be unique.")
		}
		rejected[id] = struct{}{}
	}
	return nil
}

func validateEnvelope(envelope Envelope, kind string) *specter.Error {
	if envelope.ProtocolVersion != Version {
		return &specter.Error{Code: specter.ErrProtocolVersionMismatch, Message: fmt.Sprintf("Unsupported protocol major version %d; expected %d.", envelope.ProtocolVersion, Version)}
	}
	if envelope.Kind != kind {
		return invalidMessage(fmt.Sprintf("Expected message kind %q.", kind))
	}
	if envelope.RequestID == "" {
		return invalidMessage("requestId is required.")
	}
	return nil
}

func invalidMessage(message string) *specter.Error {
	return &specter.Error{Code: specter.ErrInvalidMessage, Message: message}
}

func safeInteger(value SafeInteger) bool {
	return value >= 0 && int64(value) <= maxSafeInteger
}

func validateJSONValue(value any) bool {
	data, err := json.Marshal(value)
	if err != nil {
		return false
	}
	decoder := json.NewDecoder(strings.NewReader(string(data)))
	decoder.UseNumber()
	var normalized any
	if decoder.Decode(&normalized) != nil {
		return false
	}
	return validateDecodedJSONValue(normalized)
}

func validateDecodedJSONValue(value any) bool {
	if value == nil {
		return true
	}
	switch typed := value.(type) {
	case string, bool:
		return true
	case json.Number:
		if exact, ok := new(big.Rat).SetString(typed.String()); ok && exact.IsInt() {
			absolute := new(big.Int).Abs(exact.Num())
			return absolute.Cmp(big.NewInt(maxSafeInteger)) <= 0
		}
		float, err := typed.Float64()
		return err == nil && validateJSONFloat(float)
	case float32:
		return validateJSONFloat(float64(typed))
	case float64:
		return validateJSONFloat(typed)
	case map[string]any:
		for _, item := range typed {
			if !validateDecodedJSONValue(item) {
				return false
			}
		}
		return true
	case []any:
		for _, item := range typed {
			if !validateDecodedJSONValue(item) {
				return false
			}
		}
		return true
	}
	return false
}

func validateJSONFloat(value float64) bool {
	if math.IsInf(value, 0) || math.IsNaN(value) {
		return false
	}
	return math.Trunc(value) != value || math.Abs(value) <= float64(maxSafeInteger)
}

func uniqueNonempty(ids []string) bool {
	seen := make(map[string]struct{}, len(ids))
	for _, id := range ids {
		if id == "" {
			return false
		}
		if _, duplicate := seen[id]; duplicate {
			return false
		}
		seen[id] = struct{}{}
	}
	return true
}

func hasFields(fields map[string]json.RawMessage, names ...string) bool {
	for _, name := range names {
		if _, present := fields[name]; !present {
			return false
		}
	}
	return true
}

// SanitizeObservationError reduces a runtime failure to telemetry-safe data.
func SanitizeObservationError(err error) *StructuredError {
	var public *specter.Error
	if errors.As(err, &public) {
		if message, ok := observationErrorMessages[public.Code]; ok {
			return &StructuredError{Code: string(public.Code), Message: message}
		}
	}
	return &StructuredError{Code: string(specter.ErrInfrastructure), Message: "Runtime operation failed."}
}

var observationErrorMessages = map[specter.ErrorCode]string{
	specter.ErrCommandRejected:     "Command was rejected.",
	specter.ErrConformanceFailed:   "Runtime conformance failed.",
	specter.ErrIdempotencyConflict: "The idempotency key conflicts with an earlier Command.",
	specter.ErrInfrastructure:      "Runtime operation failed.",
	specter.ErrInvalidInput:        "Operation input is invalid.",
	specter.ErrInvalidOutput:       "Operation output is invalid.",
	specter.ErrReactionFailure:     "One or more Reactions failed.",
	specter.ErrUnknownCommand:      "Command type is not registered.",
	specter.ErrUnknownEvent:        "Event type is not registered.",
	specter.ErrUnknownQuery:        "Query type is not registered.",
	specter.ErrVersionConflict:     "Event Log version conflict.",
}
